### PR TITLE
Add notification support to markdown parser

### LINKED ISSUE #64: Implement notifications/admonitions
We should implement some tasteful design as a placeholder: at the moment Warnings etc are just styled as plain text.

### PR DESCRIPTION
**Pull request by nottrobin**
This adds the styling and the markup generation to display [Vanilla notifications](https://docs.vanillaframework.io/en/patterns/notification) in documents.

Usage:

    !!! {type} "{title}":
        {content}

This will generate notification markup, e.g.:

    <div class="p-notification--{type}">
      <p class="p-notification__response">
	<span class="p-notification__status">{title}:</span>
	<span class="p-notification__line">{content}</span>
      </p>
    </div>

You can create notifications of types [supported by Vanilla framework](https://docs.vanillaframework.io/en/patterns/notification):

- Note
- Warning
- Positive
- Negative

Note, if `"{title}"` is ommitted (`!!! Note:`), the `type` will be used instead (e.g.: `<span>Note:</span>`). If it is left blank (`!!! Note "":`) then no title will be used.

Fixes #64.

QA
---

Test this out with maas-docs:

``` bash
git clone https://github.com/canonicalltd/maas-docs  # Get maas-docs
cd maas-docs  # go to maas-docs directory
git clone -b notifications https://github.com/nottrobin/documentation-builder  # clone down my branch
python3 -m venv env3  # Get ready to install in a local environment
env3/bin/pip install ./documentation-builder  # Install my branch
env3/bin/python documentation-builder/bin/documentation-builder --force  # Build the documentation
```

Now open `build/en/index.html` and see the notifications.

You can play around with the notifications by editing `en/index.md` and building the documentation again as mentioned above.

### CODE REVIEW

**Comment by nottrobin** [comment]
[Add support for notifications](https://trello.com/c/eeXmJI5l/2841-add-support-for-notifications)

**Comment by evilnick** [comment]
before I go through this in detail I have 2 queries:

 1. Why does this implement another extension instead of using the standard admontitions one? I realise it allows us to tie in with the existing defined style, but IMO that isn't the point.

 1. As it stands, this doesn't do what we asked, which is to be able to create the standard admonitions: Info, tip, note, caution, and warning. 


welcome back :smile:

**Comment by nottrobin** [comment]
@evilnick yes the problem with the existing admonitions extension is that it doesn't give you any control over the markup that is generated. We are definitely opinionated about the structure of this markup - it needs to fit in with our existing markup patterns, so we need to write our own to generate the markup just how we want it.

Yes I am aware that this doesn't provide the standard admonitions. As I mentioned in https://github.com/CanonicalLtd/documentation-builder/issues/64#issuecomment-288069756, I'm only implementing Vanilla-style notifications at this stage as this is all we have designs for. This is what was recommended as the first step by @yaili in https://github.com/vanilla-framework/vanilla-docs-theme/issues/28.

If you need the standard admonitions instead, updating this markdown extension to support them is trivial, it would simply be a matter of updating [the class names](https://github.com/nottrobin/documentation-builder/blob/0275f56199abf22c67b964161747f2f8cae2d7b4/ubuntudesign/documentation_builder/extensions.py#L77). However, before that could happen they'd need to be designed (this will be the major delay), and then added to vanilla-docs-theme.

I recommend that we implement them this way for now so we can get it done. Since you only appear to be actually using `note` and `warning` in the existing MAAS docs, this seems like it will probably be an okay interim solution, won't it?

**Comment by evilnick** [comment]
>  We are definitely opinionated about the structure of this markup - it needs to fit in with our existing markup patterns,

I'm sure. I might have been inclined to fix the generated code with post-processing rather than deviate more from our chosen implementation, but as it uses the same syntax on the generation side I don't expect it matters so much.

> If you need the standard admonitions instead, updating this markdown extension to support them is trivial, it would simply be a matter of updating the class names

We most definitely do need the standard names for these things. It may seem faintly pedantic, but they **are** standards, and people who contribute, or indeed other non-doc team users would expect to be able to use them. Using a different set because those are the ones we have designs for seems rather tail-wagging-dog

**Comment by evilnick** [comment]
P.S. We only use Warning and note in any of our docs for exactly the same reason - we could never get anyone to design a style for any of the others so we were conditioned into not using them.

**Comment by nottrobin** [comment]
> We most definitely do need the standard names for these things. It may seem faintly pedantic, but they are standards, and people who contribute, or indeed other non-doc team users would expect to be able to use them. Using a different set because those are the ones we have designs for seems rather tail-wagging-dog

I hear you. It does seem like the tail wagging the dog.

But what it comes down to is you can have Vanilla notifications now, and then we can implement DocBook admonitions in a couple of months after all the predicate discussions have been had. Or you can have nothing now, and have DocBook admonitions in a couple of months when all the predicate discussions have been had.

**Comment by evilnick** [comment]
Sure, I am definitely in favour of having something now, hence the original issue wherever it was :smile: 
I just don't want anyone to imagine that this is going to stop me mentioning it repeatedly. 

I am testing it now...

File: ubuntudesign/documentation_builder/extensions.py [hunk 1]
```diff
@@ -0,0 +1,157 @@
...
+                '    {% endif %}' +
+                '    <span class="p-notification__line">{{body}}</span>' +
+                '  </p>' +
+                '</div>'
+            )
+            notification_type, title = self.get_type_and_title(match)
+
+            markup = template.render(
+                {
+                    'class': self.type_classes[notification_type],
```
Comment (evilnick) [inline ubuntudesign/documentation_builder/extensions.py]: My only comment here is do we want to have a fallback default class (note?) - at the moment any non-matching type throws an error.

**Comment by evilnick** [comment]
The lines of colour generated by warnings are such a welcome relief that I fear we may invent warnings just to make the pages look nicer

http://phoenixrising.me/treating-cfs-chronic-fatigue-syndrome-me/living-ii-mindfulness/chronic-fatigue-syndrome-mecfs-treatments-the-stress-busters-aromatherapy

File: ubuntudesign/documentation_builder/extensions.py (hunk 1 above)
Comment (nottrobin) [inline ubuntudesign/documentation_builder/extensions.py]: Ah! Good idea. Then you could start using Admonition types now (`!!! Tip:`), but just have them display as normal notes. That'd be handy. I'll do that now.

**Comment by nottrobin** [comment]
@evilnick I've updated the way that class is created to any unrecognised type of notification will appear as the default one. E.g., you could do:

```
!!! Tip:
    Here's a tip
```

I'll merge this now.

Please continue the discussion about the design requirements for Admonitions, as well as discussions about the greyscale fatigue issue, over in https://github.com/ubuntudesign/vanilla-docs-theme-design/issues/3.

**Pull request merged**